integration/aggregate_compat_test.go

```diff
@@ -0,0 +1,350 @@
+
+			var nonEmptyResults bool
+
+			t.Run(targetCollection.Name(), func(t *testing.T) {
+				t.Helper()
+
+				var targetRes, compatRes []bson.D
+				targetErr := targetCollection.Database().RunCommand(ctx, command).Decode(&targetRes)
+				compatErr := compatCollection.Database().RunCommand(ctx, command).Decode(&compatRes)
+
+				if targetErr != nil {
+					t.Logf("Target error: %v", targetErr)
+					t.Logf("Compat error: %v", compatErr)
+					AssertMatchesCommandError(t, compatErr, targetErr)
+
+					return
+				}
+				require.NoError(t, compatErr, "compat error; target returned no error")
+
+				AssertEqualDocumentsSlice(t, compatRes, targetRes)
+
+				if len(targetRes) > 0 || len(compatRes) > 0 {
+					nonEmptyResults = true
+				}
+			})
+
+			switch tc.resultType {
+			case nonEmptyResult:
+				assert.True(t, nonEmptyResults, "expected non-empty results")
+			case emptyResult:
+				assert.False(t, nonEmptyResults, "expected empty results")
+			default:
+				t.Fatalf("unknown result type %v", tc.resultType)
+			}
+		})
+	}
+}
+
+func TestAggregatePipelineCompat(t *testing.T) {
+	testCases := map[string]aggregatePipelineCompatTestCase{
+		"CollectionAgnostic": {
+			command: bson.D{
+				{"aggregate", 1},
+			},
+			skip: "https://github.com/FerretDB/FerretDB/issues/1890",
+		},
+		"FailedToParse": {
+			command: bson.D{
+				{"aggregate", 2},
+			},
+			resultType: emptyResult,
+		},
+		"PipelineTypeMismatch": {
+			command: bson.D{
+				{"aggregate", "collection-name"},
+				{"pipeline", 1},
+			},
+			resultType: emptyResult,
+		},
+		"StageTypeMismatch": {
+			command: bson.D{
+				{"aggregate", "collection-name"},
+				{"pipeline", bson.A{1}},
+			},
+			resultType: emptyResult,
+		},
+	}
+
+	testAggregatePipelineCompat(t, testCases)
+}
+
+func TestAggregateCompatCount(t *testing.T) {
+	testCases := map[string]aggregateStageCompatTestCase{
+		"Value": {
+			pipeline: bson.A{bson.D{{"$count", "v"}}},
+		},
+		"NonExistent": {
+			pipeline: bson.A{bson.D{{"$count", "nonexistent"}}},
+		},
+		"Location15948": {
+			pipeline:   bson.A{bson.D{{"$count", "_id"}}},
+			resultType: emptyResult,
+		},
+		"Location40156": {
+			pipeline:   bson.A{bson.D{{"$count", 1}}},
+			resultType: emptyResult,
+		},
+		"Location40157": {
+			pipeline:   bson.A{bson.D{{"$count", ""}}},
+			resultType: emptyResult,
+		},
+		"Location40160": {
+			pipeline:   bson.A{bson.D{{"$count", "v.foo"}}},
+			resultType: emptyResult,
+		},
+		"Location40158": {
+			pipeline:   bson.A{bson.D{{"$count", "$foo"}}},
+			resultType: emptyResult,
+		},
+	}
+
+	testAggregateStageCompat(t, testCases)
+}
+
+func TestAggregateCompatMatch(t *testing.T) {
+	testCases := map[string]aggregateStageCompatTestCase{
+		"ID": {
+			pipeline: bson.A{bson.D{{"$match", bson.D{{"_id", "string"}}}}},
+		},
+		"Int": {
+			pipeline: bson.A{
+				bson.D{{"$match", bson.D{{"v", 42}}}},
+				// sort by _id because compat does not have deterministic order.
+				bson.D{{"$sort", bson.D{{"_id", 1}}}},
```

Comment:
We should do it in `testAggregateStageCompat` itself; see

https://github.com/FerretDB/FerretDB/blob/852ebf4002669578fd332c90d5aff8f9af08a66f/integration/query_compat_test.go#L67-L70

There we should not use a separate field, but we could check if `pipeline` includes `$sort`

Reply:
Added check to find $sort in the pipeline, and adding $sort if that doesn't exist :+1: